### PR TITLE
ColorPalette, GradientPicker: fix color picker popover positioning

### PR DESCRIPTION
**Pull request by ciampo**
## What?

This PR aims at improving the positioning of the color picker's popover in all color-related widgets (mostly `ColorPalette`, and `CustomGradientBar` for gradients) after #40740 introduced a regression that stopped the forwarding of `popoverProps` correctly in the `ColorPalette` component (see [change](https://github.com/WordPress/gutenberg/pull/40740/files#diff-b53300a311404baec5a0c0cfc1d898fa89050f29959680b6317243be716c3691)).

This PR also slightly tweaks the positioning of the popover for all cases.

## Why?

With this change, the `GradientPicker` component can override `popoverProps` again as it was previously intended.

## How?

 - Take into account the received `popoverProps` in the `CustomColorPickerDropdown` component
 - Popover positioning logic:
   - Make sure that the color picker's popover in gradient bar always opens to the `bottom` of the gradient's control point (centred on the x axis)
   - For all other cases:
     - If the popover is rendered in the sidebar, open the popover to the left of its anchor (this causes the popover to "stack" to the left in case its anchor is another popover)
     - if the popover is not rendered in the sidebar, open the popover to the `bottom` of its anchor (centred on the x axis), like for the gradient control points.
 - Make the popover always `shift`, meaning it can't render outside of the viewport
 - Use the `placement` property, instead of the legacy `position`
 - Bonus: use the `HStack` component instead of custom CSS styles to minimize the amount of style overrides

## Testing Instructions

In Storybook, verify that the popover is always rendered inside the viewport (i.e. shifting behavior is enabled)

In the site editor, try to edit the available gradients palette and focus on the popover positioning for each color stops

## Screenshots or screencast 

_Note: the following screenshots / videos may be out of date. Please refer to the latest comments in this PR for the final popover behavior_

`GradientPicker` in Storybook, `trunk`:

https://user-images.githubusercontent.com/1083581/182940241-1dfee265-c54c-4f59-87af-e77551dd60d9.mp4

`GradientPicker` in Storybook, this PR:

https://user-images.githubusercontent.com/1083581/182940361-509e22c8-4fd1-4fd3-8dbc-b8d904f908c4.mp4

Popover for color stops in gradient picker, site editor, `trunk`: 

https://user-images.githubusercontent.com/1083581/182940472-da0d08ef-dd25-4cab-bd48-1df951fc29ef.mp4

Popover for color stops in gradient picker, site editor,  this PR:

https://user-images.githubusercontent.com/1083581/182940489-92e137c8-8565-49c7-babc-59fdb0c3ad2c.mp4

(@jasmussen , welcome back! I may need some guidance here about what is the expected popover positioning when editing color stops in the site editor — see video above)

### CODE REVIEW

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (ciampo) [inline packages/components/src/custom-gradient-bar/control-points.js:88]: We can pass the entire `ref` object here, no need to pass the `current` value

File: packages/components/src/color-palette/index.js
Comment (ciampo) [inline packages/components/src/color-palette/index.js]: @talldan , do you think that the offset / positioning are working as expected here (when editing gradients in global styles in the site editor) ? In my tests, it felt like the popover is not showing the behavior that I expected

Comment (talldan) [reply packages/components/src/color-palette/index.js]: What are you seeing? Something that tripped me up initially is that `Dropdown` has a default offset of `13`, so `20` is only offsetting by `7` from the default position.

The popover is using the `components-custom-gradient-picker__markers-container` element as the anchor/reference element, so the positioning looks correct (checked by passing `0` for the offset prop). I'm not sure how it ends up with that as the reference element.

I tried `200` for an offset and it was about 200 pixels below the reference, which is to be expected with the `bottom` placement—the main axis will be the direction of offset.

**Comment by talldan** [comment]
BTW - I notice in WordPress core, the custom color popover is always to the left of the gradient popover (it seems anchored to the popover contents, judging by how it sits just below the top padding, and I think it would be placement `left-start`):
Screen Shot 2022-08-05 at 1 29 48 pm

Maybe that's the desired behavior? Hopefully Joen can confirm.

**Comment by jasmussen** [comment]
Thank you for the welcome! This is a tricky one. Here's a GIF showing the desired test behavior:

taller

The color picker opens downward and to the right from the opening swatch (cropped by the GIF).

This is in comparison to trunk, where the picker seems arbitrarily positioned:

trunk

The desired behavior in the past was for each popover to open in a stacking behavior leftwards from the ItemGroup button — swatch popover first, then the color popover left of that. A bit like this inspector mockup where I hacked the CSS:

Screenshot 2022-08-05 at 10 05 19

My opinion: that hasn't been entirely successful.

* The color popover has a unique borderless and frameless design that, when stacked, looks a little different
* The color context is confused, relying on an active state for the gradient
* Povers opening from popovers feels like too much

An older mockup, is to enable the swatch panel to _slide_ rightwards to the color picker:

Color ItemGroup Flow i5

It solves the popover in popover problem, but it's still not clear that would be a better user experience, moreover, it's likely going to be a lot of work. So outside of a larger effort, we can either land this PR and drop the stacking behavior, or see if we can restore it to its original behavior (the inspector mockup above). 

Question: is there a substantial code quality improvement in dropping the stacking behavior? I imagine there might be, and if we do go that way, I'd prefer prefer the popover open down and center, rather than down and right. Like so:

Screenshot 2022-08-05 at 10 14 37

I'm keen on other opinions.

File: packages/components/src/color-palette/index.js
Comment (ciampo) [reply packages/components/src/color-palette/index.js]: > Something that tripped me up initially is that `Dropdown` has a default offset of `13`, so `20` is only offsetting by `7` from the default position.

Woops, that was totally it 🤦  nevermind me!

File: packages/components/src/color-palette/test/__snapshots__/index.js.snap
Comment (ciampo) [inline packages/components/src/color-palette/test/__snapshots__/index.js.snap:360]: Just noting that this snapshot change is due to a change in `CustomColorPickerDropdown` (in `packages/components/src/color-palette/index.js`), where previously the `__unstableShift` prop was set only when `isRenderedInSidebar` was `true` — while, with the changes in this PR, `__unstableShift` is _always_ set to `true`

**Comment by ciampo** [comment]
Hey @jasmussen , I made some changes, and the popover should now open below (and centred) under each control point in the gradient bar, if I understood this piece of feedback correctly:

>  I'd prefer prefer the popover open down and center, rather than down and right. Like so:
> 
> Screenshot 2022-08-05 at 10 14 37


Otherwise, I'm also happy to try and replicate what you called the "original behavior":

> The desired behavior in the past was for each popover to open in a stacking behavior leftwards from the ItemGroup button — swatch popover first, then the color popover left of that. A bit like this inspector mockup where I hacked the CSS:
> 
> Screenshot 2022-08-05 at 10 05 19

With respect to a deeper refactor, I personally never liked having "nested" popovers, and so I'd be in favour of having something similar to the older mockup that you shared, where the swatch panel slides rightwards to the color picker (although, as you guessed, that would definitely be more than a simple refactor).

File: packages/components/src/color-palette/index.js
Comment (talldan) [inline packages/components/src/color-palette/index.js]:
```suggestion
		// Open below the control point (centered aligned) when in the sidebar.
```

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (talldan) [inline packages/components/src/custom-gradient-bar/control-points.js]: I realise it's a bit like this in trunk, but this line makes the duotone color picker look like this, overlapping the block toolbar:
Screen Shot 2022-08-09 at 12 50 18 pm

It might be good to simplify it and have `GradientColorPickerDropdown` always positioned underneath the control point. When there's not enough space `Popover` will already flip it to the `top`.

Would be good to get more thoughts on this from @jasmussen and possibly @ajlende.

Another note is that the `useMemo` here could be removed, as `CustomColorPickerDropdown` doesn't respect it and creates a new object on every render.

File: packages/components/src/color-palette/index.js
Comment (talldan) [inline packages/components/src/color-palette/index.js]: This change makes the CustomColorPickerDropdown for normal colors (non-gradients) appear differently.

Trunk:
Screen Shot 2022-08-09 at 1 11 16 pm

This branch:
Screen Shot 2022-08-09 at 1 11 44 pm

I can't see that discussed so far, the conversation has mostly been about the gradient picker.

@jasmussen Does this look ok, or should it stay to the left of the color palette dropdown?

File: packages/components/src/color-palette/index.js
Comment (talldan) [inline packages/components/src/color-palette/index.js:116]: Not introduced here, but this is the first time I've seen it. `isRenderedInSidebar` doesn't seem like it should be in the components package. The prop is very specific to the editor's right-hand sidebar. What if there's a sidebar on the left of the screen?

It looks like it was a late fix for a WordPress release - https://github.com/WordPress/gutenberg/pull/37115. @jorgefilipecosta, as you introduced this would you be willing to revisit it?

My feeling is that it should be more generic, like a `childPopoverPlacement` prop that's passed through to any popovers that should cascade in a given direction.

**Review by talldan — COMMENTED** [review]
This looks pretty close. I noticed a couple of things that might need some more input from a designer.

**Comment by jasmussen** [comment]
Thank you for the fast and impressive work, and thank you Dan for the very helpful pointers towards the changes beyond just gradients.

For the gradients specifically, they do feel more contextual with the popover:
Screenshot 2022-08-09 at 09 25 43

I failed to think about the other places where this component is used, and per Dan's pointers, they aren't optimal:

Screenshot 2022-08-09 at 09 25 59

(Yes Duotone should probably always open control points downwards)

Screenshot 2022-08-09 at 09 25 35

Specifically this one, clicking the main color swatch to customize it, feels confusing. That does make me feel like we should seek back to the original behavior. Sorry about the detour.

On a separate track with less urgency, yes, we should still think about how we can improve this flow overall. Whether that's the sliding interface or something else — this PR has helped put into clarity what needs to be considered here. Thanks again.

File: packages/components/src/color-palette/index.js
Comment (ciampo) [reply packages/components/src/color-palette/index.js:116]: I agree 100% on this one, and coincidentally I have just opened an issue with a few changes that I'd like to see for `GradientPicker`:

- https://github.com/WordPress/gutenberg/issues/42994

> My feeling is that it should be more generic, like a `childPopoverPlacement` prop that's passed through to any popovers that should cascade in a given direction.

For the specific prop, it will largely depend on what we decide design-wise — let's see how this conversation evolves — but in general, I'd _love_ to get rid of that `isRenderedInSidebar` prop (also considering how many times it's forwarded through the various component chain!)

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (ciampo) [reply packages/components/src/custom-gradient-bar/control-points.js]: Just pushed https://github.com/WordPress/gutenberg/pull/42989/commits/1d30a2a52808ba02ce90780cde8132419dec7c40 in which:

- the `placement` is always set to `bottom` regardless of the `isInSidebar` prop (I also tweaked the `offset` by a couple of px)
- I've used `useMemo` also inside `CustomColorPickerDropdown` (instead of removing it)

File: packages/components/src/color-palette/index.js
Comment (ciampo) [reply packages/components/src/color-palette/index.js]: Since joen replied to this question in [his comment](https://github.com/WordPress/gutenberg/pull/42989#issuecomment-1209014992) below, I'll mark this convo as closed and will continue the conversation by replying directly to his comment

**Comment by ciampo** [comment]
Alright, I've applied _some_ of the suggested changes: in particular, with the latest changes in this PR _all dropdowns_ opening within a `GradientPicker` or a `ColorPalette` currently open below the clicked anchor, regardless of whether they are rendered in the sidebar or not.

This is the thing: with the new desired behavior, the `__experimentalIsInSidebar` prop is not useful anymore to determine which configuration (placement/offset...) we should apply to the popover — for example, the scenario described by @jasmussen above also happens in the sidebar:

> Screenshot 2022-08-09 at 09 25 35
> 
> Specifically this one, clicking the main color swatch to customize it, feels confusing. That does make me feel like we should seek back to the original behavior. Sorry about the detour.

This also links well with @talldan 's [comment](https://github.com/WordPress/gutenberg/pull/42989#discussion_r940919179) above. From what I gathered, in order to achieve the new desired behavior we'd need to:

- remove the `__experimentalIsInSidebar` prop from the `ColorPalette`, `CustomGradientBar`, `CustomGradientPicker`, `GradientPicker`, `PaletteEdit` components, and from all of the consumers of these components (since it's not useful anymore)
- expose a new prop on these components that will allow us to customize the position of the popover — the most flexible way to achieve that is to literally accept a `childPopoverProps` or `colorPickerPopoverProps` prop, where we can set any desired popover behavior

This is definitely a much larger job than the current size of the PR, and therefore I'll wait for everyone's feedback on the proposed solution before working on it.

It would be great if we could avoid the many layers of prop drilling the currently are needed to forward the `__experimentalIsInSidebar` prop all the way down to the low-level components, and so if anyone has a better alternative to adding a new prop, that would be great.

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (ciampo) [reply packages/components/src/custom-gradient-bar/control-points.js]: Actually, after https://github.com/WordPress/gutenberg/pull/42989/commits/7c2352845d08edfac00d31e0b47f16a049f03e80, there's currently no need for any `popoverProps` override (although there will probably be a need for it again soon)

**Comment by jasmussen** [comment]
Just to take stock of the current status. Here's what trunk looks like:
trunk

Here's what this branch looks like:

opener

As discussed on the thread, there's a lot to fix in terms of popover appearances and positions, though I did note a slight preference for opening like trunk does at least for _colors_. However as is also clear, the bug to be fixed is the one for color stops in the gradient. 

In order to fix that, is it fine that we open down/center from the opening color stops across the board? Probably, yes. Especially so we can land a fix for 6.1. I'd love a quick input from @javierarce or @jameskoster if they have alternative ideas, though.

One thing we'd need to fix if we go with this approach, though, is that the animation indicates a top left origin, and it should indicate a top center origin. It's hard to see with the fast animation, but if you look for it, it's there. Nice work.

**Comment by jameskoster** [comment]
I've always found it a bit confusing that we combine presets and custom values in the same panel. We could eliminate the double popovers altogether if we did something like:

Screenshot 2022-08-10 at 10 11 19

I would probably combine preset solids and gradients too, but that's another story.

Anyways...

> In order to fix that, is it fine that we open down/center from the opening color stops across the board

The overlapping popovers feels a bit awkward for solid colors, but not necessarily a _regression_. It feels like a fair trade-off to fix the original bug.

**Comment by ciampo** [comment]
I will also try to summarize the current situation, in case we can get to a simpler solution for the purpose of this PR (which originally aimed at being a "quick" bug fix):

 - Currently on `trunk`, when we display a `ColorPicker` inside a `Popover`, we have two "configurations" that are chosen based on whether the popover is rendered in the sidebar or not:
   - If it's rendered in the sidebar, the popover's placement is supposed to be `left-start` (i.e. to the left, and with the top edge of the popover aligned with the top edge of the anchor) with respect to its anchor
   - If it's rendered outside of the sidebar, the popover is supposed to render above its anchor
   - The anchor depends on each "widget" and can also change based on the `isRenderedInSidebar` check: sometimes the anchor is the individual "circle" (single color entries in the color palette, or control points in a gradient picker), sometimes it's the larger "rectangle" (the whole clickable color indicator, or the whole gradient bar)
 - **The limit with this system is that, as explained above, we are currently able to only pick between 2 "configurations" based on the `isRenderedInSidebar` check**. And so, if we want the popover to be configured the same way as when in the sidebar also when it's used by the Duotone tools _outside_ of the sidebar, it means that:
   - We either have to use the same "configuration" for _all_ color picker popovers
   - We need to spend some time refactoring the code to allow for a finer grain control of the popover configuration that is more flexible than a simple `isRenderedInSidebar` check

Currently, the behaviour of the color picker popover in  `trunk` can also be a bit weird and inconsistent because of the regressions introduced by #40740 (which we're working on fixing). For the sake of keeping this PR focused on fixing the initial bug, we could also find a solution that still relies on having two configurations that we switch based on the `isRenderedInSidebar` check. We could aim at restoring the original intended behavior, ie.:

- if the popover is in the sidebar, we do what @jasmussen referred to as the "original desired behavior": _each popover to open in a stacking behavior leftwards_
- if the popover is not in the sidebar (including duotone), we pick a placement for the popover (which used to be on `top` of the anchor, if there's enough space)

**Comment by talldan** [comment]
> The limit with this system is that, as explained above, we are currently able to only pick between 2 "configurations" based on the isRenderedInSidebar check.

I think there's the possibility of three configurations:
- The `GradientColorPickerDropdown` version of `CustomColorPickerDropdown` can have its own behavior by overriding the `propoverProps`. Currently it seems like the discussion has leaned towards always opening the popover below the control points.
- Standard versions of `CustomColorPickerDropdown` that don't have an override of `popoverProps` can have two behaviors depending on `isRenderedInSidebar`. The discussion has suggested opening to the left of the parent popover when `isRenderedInSidebar` is `true`, but I'm not sure how it should appear when `isRenderedInSidebar` is `false`. Below?

**Comment by ciampo** [comment]
@talldan's proposed solution makes sense to me — I went ahead and tried to implement it with my latest changes.

Here's what the "color picker in a popover" behaves like across the editor:

https://user-images.githubusercontent.com/1083581/184196225-983e37aa-19d8-4258-bb97-4b85b26ecc80.mp4


What do folks think?

**Comment by jasmussen** [comment]
Thanks for your effort. If the code isn't too painful, I think that's an excellent path forward, even if we know we want to revisit this in the future. 

If you can, I'd love to put 12px of space between the color picker and the swatch palette, though — just like the space between the ItemGroup item and the swatch palette. And the color picker still opens from top left, as opposed to top center, for the gradient stops.

Nice work!

**Comment by ciampo** [comment]
Hey @jasmussen , I've updated the offset between the color picker and the swatch palette:

Screenshot 2022-08-12 at 10 01 23

>  And the color picker still opens from top left, as opposed to top center, for the gradient stops.

I will work on improving the popover opening animations in a follow-up PR.

I believe this PR is ready for a final review and hopefully we can go ahead and merge it!

**Comment by jasmussen** [comment]
Screenshot looks good to me (maybe give or take some border math, is it slightly larger than the space between palette and itemgroup?) but nothing to block it.

I tried compiling the branch but for whatever reason I'm still seeing the old spacing. Probably something on my side. But in any case, screenshot looks good!

**Comment by ciampo** [comment]
> (maybe give or take some border math, is it slightly larger than the space between palette and itemgroup?) but nothing to block it.

It could be — nothing that we can't tweak in a follow-up if necessary (in case you wanted to test it on your machine, [these would be the lines of code](https://github.com/WordPress/gutenberg/blob/f731cf2e17ff906ece2304818d46522e61403a23/packages/components/src/color-palette/index.js#L125-L128) responsible for positioning that popover)

Do you think this PR is good enough to get an approval and merge?

**Review by jasmussen — APPROVED** [review]

**Review by talldan — APPROVED** [review]
LGTM! Great work iterating on this one.

**Comment by ciampo** [comment]
> I will work on improving the popover opening animations in a follow-up PR.

Opened https://github.com/WordPress/gutenberg/pull/43186 (I had already started work on this https://github.com/WordPress/gutenberg/pull/42531 and so it was quite quick to put the PR together)

File: packages/components/src/color-palette/index.js
Comment (jorgefilipecosta) [reply packages/components/src/color-palette/index.js:116]: I agree it would be good to get rid of isRenderedInSidebar it was added as a short fix for a visual issue.

Comment (ciampo) [reply packages/components/src/color-palette/index.js:116]: @jorgefilipecosta , would you have any capacity to help with the work on this component, also since you are its main author ?

Comment (jorgefilipecosta) [reply packages/components/src/color-palette/index.js:116]: Hi @ciampo, I will try to find some availability to work on some enhancements to this component.

Comment (ciampo) [reply packages/components/src/color-palette/index.js:116]: That's great, thank you! I've assigned [the issue with the list of enhancements ](https://github.com/WordPress/gutenberg/issues/42994) — feel free to ping for any help and for reviews :)

**Pull request merged**
